Patch/clarify webhook

robertjahn opened the pull request.

robertjahn commented:
This will also address https://github.com/keptn/keptn.github.io/issues/1037

oleg-nenashev reviewed (commented):
The addition to the main page looks nice! Just a style thing, maybe it makes sense to split long paragraphs to multiple lines like it was in the original content. It is useful for code reviews and for incremental pull requests.

I also wonder about removing the Jenkins webhook example. It seems to be a nice addition if this is still relevant. Should it be rather moved to Tutorials and referenced from there? Same for Slack that it rather on ArtifactHub now

oleg-nenashev reviewed (changes requested):
Requesting changes to indicate there is a discussion to be done @robertjahn

agardnerIT commented:
More than happy to be overruled but I have a problem with the `interactive` and `non interactive` webhook terminology. The first time I heard those terms I thought **I** had to interact to do something.

Perhaps add some text like: "If you would like a third party too to **act**, you most likely want to subscribe to the `.triggered` event to trigger your external tool. In this case, the external tool is responsible for sending a `.started` and `.finished` event back to Keptn.
If you want to notify a tool that something has occurred (for example notify a messaging platform) , you'd usually subscribe to a `.finished` event. In this case, the integration is fire-and-forget so you don't need to send any other events."

grabnerandi commented:
Chiming in on the discussion regarding interactive and non-interactive. How about "active" and "passive" integrations? Active means that the integration actively applies some changes to the environment, e.g: deploy, execute a test, ... -> passive on the other hand is just consuming the data without changing the impact of the sequence flow.
besides active & passive you could also use "Actor" & "Consumer"

Just my thoughts as a non-native speaker that also thinks that interactive and non-interactive sounds a bit weird

robertjahn commented:
I agree the name should be easy to refer to and clear what is means.  Some options.  I am leaning towards the one with stars because they do describe better how they work, but they are kinda long in the name.

ONEWAY
a) Asynchronous
b) unidirectional
c) fire-and-forget *
d) silent
e) non-interactive
f) passive
g) broadcast

TWOWAY
a) Synchronous
b) request-response
c) Synchronous request-response *
d) bidirectional
e) interactive
f) active

here is a good references I found for pattern types: https://docs.oracle.com/cd/E23943_01/doc.1111/e17363/chapter05.htm#FPCON244 
https://www.enterpriseintegrationpatterns.com/patterns/conversation/BasicIntro.html

miigwi commented:
I like the "active" / "passive" differentiation for the webhook integrations. Gives a better picture than "interactive" and "non-interactive".

johannes-b commented:
I like the idea @grabnerandi was proposing: differentiation between *active* and *passive*. 
These two terms are short, precise, and bring it to the point.

robertjahn commented:
ok seem to be consensus on active and passive.  I have adjusted accordingly in my last commit

johannes-b commented:
@robertjahn please do not delete: content/docs/0.13.x/integrations/webhooks/jenkins/index.md 
We still need it since there is no Jenkins example on the ArtifactHub for Jenkins: 
![image](https://user-images.githubusercontent.com/729071/155736703-a8ff4ea7-a39b-4a9e-95a9-8781f547018a.png)

johannes-b commented on content/docs/0.13.x/integrations/webhooks/_index.md:
```diff
@@ -5,74 +5,96 @@ weight: 2
+Webhooks allow you to easily integrate various third-party tools such as testing services, CI/CD pipelines, and incident management services. In addition to just invoking the third-party webhook receiver, webhook subscriptions can be configured to wait for the third-party webhook receiver to send back details about what it did. 
 
-## Create a Webhook integration
+Depending on the receiving tools configuration requirements, data can be either passed within the payload body or as query params in the URL along with any required headers and API tokens.
```
Should be: "... receiving tool's configuration ..."

johannes-b commented on content/docs/0.13.x/integrations/webhooks/_index.md:
```diff
@@ -5,74 +5,96 @@ weight: 2
-Webhooks are created at a *Task* level and can be triggered by the following [event types](https://github.com/keptn/spec/blob/0.2.2/cloudevents.md#task-events):
+## Webhook subscriptions
+
+Webhook configurations, referred to as subscriptions, are created at a *Task* level and are configured to listen to one of the the following task [event types](https://github.com/keptn/spec/blob/0.2.2/cloudevents.md#task-events):
```
Remove: ~~the~~ the following

johannes-b commented on content/docs/0.13.x/integrations/webhooks/_index.md:
```diff
@@ -5,74 +5,96 @@ weight: 2
-* *Filter*: To restrict the webhook to certain stages and services you can specify those using filters.
-
-*Note:* Custom data in the event payload is available only for `triggered` task suffix.
+* *Task suffix*: The state of the task when the webhook should be fired; select one of: `started`, of `finished`
```
`, of finished ` should be  `, or finished`

johannes-b commented on content/docs/0.13.x/integrations/webhooks/_index.md:
```diff
@@ -82,112 +104,194 @@ Based on the Go templating capabilities, you can:
+**Step 4: Set "Send Finished Event" flag**
 
-## Include sensitive data
+The `Send finished` event flag is not applicable to a *non-interative* webhook. So you can leave it unselected as shown below.
```
Please change `non-interactive` to `passive`

johannes-b commented on content/docs/0.13.x/integrations/webhooks/_index.md:
```diff
@@ -82,112 +104,194 @@ Based on the Go templating capabilities, you can:
+    width="400px">}}
+1. Click the **Create subscription** button to save and enable the webhook for your integration.
+
+**Step 2: third-party webhook receiver**
```
third-party > Third-party

johannes-b commented on content/docs/0.13.x/integrations/webhooks/_index.md:
```diff
@@ -82,112 +104,194 @@ Based on the Go templating capabilities, you can:
-            name: api
-            key: api-token
+1. Follow the same steps above to add a new webhook subscription, but configure it to listen to a Task suffix of `triggered`. 
+1. Ensure that that the `Automatically` option is selected as shown below. When this is set, the webhook service will automatically send the task `finished` event for your subscription.
```
Ensure ~~that~~ that

johannes-b commented on content/docs/0.13.x/integrations/webhooks/_index.md:
```diff
@@ -82,112 +104,194 @@ Based on the Go templating capabilities, you can:
+*Important Note:* 
 
-For more advanced configuration options, you can modify the raw request declared as [curl](https://curl.se/) command. Therefore, you need to access the `webhook.yaml` config file in the Git repo you set for an upstream. In this Git repo, you find the `webhook.yaml` file based on the filters you selected on the task subscription, e.g., if a filter is set for stage *production*, go to the *producti ...
+* If the third-party webhook receiver does NOT send back the task `finished` event, the sequence will remain in a waiting states and will not continue until it times out or the it receives the task `finished` event.
```
.. in a waiting states ... > in a waiting state

johannes-b replied:
... or ~~the~~ it receives the task ...

johannes-b commented on content/docs/0.13.x/integrations/webhooks/_index.md:
```diff
@@ -82,112 +104,194 @@ Based on the Go templating capabilities, you can:
+  * `{{.shkeptncontext}}` - required in the `shkeptncontext` attribute of the `finished` event
+  * Optional, but depends on your implementation: `{{.data.project}}`, `{{.data.service}}`, `{{.data.stage}}`
+1. Follow the same steps above to add a new webhook subscription, but configure it to listen to a Task suffix of `triggered`. 
+1. Ensure that that the `By webhook receiver` option is selected as shown below. When this is set, the webhook service will expect the webhook receiver to send back the task `finished` event.
```
Ensure ~~that~~ that

johannes-b commented on content/docs/0.13.x/integrations/webhooks/_index.md:
```diff
@@ -82,112 +104,194 @@ Based on the Go templating capabilities, you can:
+
+The third-party webhook receiver can also send back custom information within the `finished` event. 
+
+This custom data will be automatically appended to every remaining `triggered` task event in the sequence.  The data much be passed in a attribute with the name of the task.  
```
.. passed in a**n** attribute ...

robertjahn commented:
> @robertjahn please do not delete: content/docs/0.13.x/integrations/webhooks/jenkins/index.md We still need it since there is no Jenkins example on the ArtifactHub for Jenkins: ![image](https://user-images.githubusercontent.com/729071/155736703-a8ff4ea7-a39b-4a9e-95a9-8781f547018a.png)

@johannes-b  -- I put Jenkins back.  But I removed the slack page.  But to make jenkins not so lonely at the top of the page, I added an examples subpage.  
<img width="749" alt="image" src="https://user-images.githubusercontent.com/24907089/155754400-f503c631-0a33-478f-8266-89fcbea984cc.png">

johannes-b reviewed (approved):
lgtm

oleg-nenashev reviewed (approved):
LGTM, thanks for the patches!
